README.md
#  RESTful Drones (restfuldrones)


This is a code repository for implement a modern and scalable RESTful Web Service with Django and the Django REST framework. This repository is in a continuous development stage, so be sure to follow the evolution of this repository. Please check the CHANGELOG file to see the new in this project.

We will code a RESTful Web Service that will allow us to register competitions for drones that are grouped into
drone categories. we want to be able to easily retrieve all the drones that belong to a specific drone category. Thus, we will have a relationship between a drone and a drone category.

CHANGELOG
# Change Log
All notable changes to this project will be documented in this file.
 
The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## [0.0.5] - 2020-02-13

### Added
- Add pytest.ini file.

### Changed
- Update drones/views.py: Add throttling policies for the class-based views related
to pilots, drones.
- Update drones/tests.py file: Add unit tests.
- Update restfuldrones/setting.py file: Add global config to default throttling classes and the desired rates.

### Fixed
- drones/views.py file. Rename 'filter_fields' to 'filterset_fields' attribute to fixed warning generated during unit tests with Pytest: MigrationNotice: `DroneCategoryList.filter_fields`
attribute should be renamed `filterset_fields`. [See:](https://django-filter.readthedocs.io/en/master/guide/migration.html#view-attributes-renamed-867)


## [0.0.4] - 2020-02-12

### Added
- Add drones/custompermission.py file: Added functionality to verify if the request of the HTTP verb is a secure method
and we assign grants permissions to the request.
- Add drones/migrations/0003_drone_owner.py: The drones_drone table was updated by adding the owner field and the admin
user was assigned as the owner of the previously created drones.

### Changed
- Update restfuldrones/urls.py file: Enable url for login user.
- Update drones/views.py: The perform_create method was overwritten. Now, when a new Drone is created, the user
associated with the request will be saved as its owner, the functionality for permission policies was also created,
and token-based authentication was enabled.
- Update drones/serializer.py: Support functionality was added to serialize drones related to a user and declare an
instance of the UserDroneSerializer class, in addition to this update you can see the user name of the owner of the
created resource..
- Update drones/models.py file: Add authentication to model Drone.
- Update restfuldrones/setting.py file: Functionality was added to some classes to use user authentication.


## [0.0.3] - 2020-02-10

### Added
- Add drones/custompagination.py file: To overrides the value of maximum_limit in pagination,

### Changed
- Update drones/views.py: Add filter, search and sort functionality to the name field.
- Update drones/models.file: Add constraint to the name fields of the Pilot, Drone and DroneCategory models so that these are unique.
- Update restfuldrones/setting.py file: Add global configuration to enable custom pagination and the django_filters application to enable filtering, searching, and ordering
features.


## [0.0.2] - 2020-02-08

### Added
- Add drones/urls.py file: Defines the URL patterns,

### Changed
- Update drones/view.py file: Some class-based views have been created, along with generic classes combined with
serializer classes.
- Update restfuldrones/urls.py file: uncomment and enable urlpatterns to include the URL patterns declared.
in the previously coded drones/urls.py file.


## [0.0.1] - 2020-02-06

### Added

- Initial folders and files to begin to build a Django RESTful Web Service project called "restfuldrones".


drones/migrations/0001_initial.py
# Generated by Django 2.2.5 on 2020-02-06 05:08

from django.db import migrations, models
import django.db.models.deletion


class Migration(migrations.Migration):

    initial = True

    dependencies = [
    ]

    operations = [
        migrations.CreateModel(
            name='DroneCategory',
            fields=[
                ('id', models.AutoField(auto_created=True, primary_key=True, serialize=False, verbose_name='ID')),
                ('name', models.CharField(max_length=250)),
            ],
            options={
                'ordering': ('name',),
            },
        ),
        migrations.CreateModel(
            name='Pilot',
            fields=[
                ('id', models.AutoField(auto_created=True, primary_key=True, serialize=False, verbose_name='ID')),
                ('name', models.CharField(default='', max_length=150)),
                ('gender', models.CharField(choices=[('M', 'Male'), ('F', 'Female')], default='M', max_length=2)),
                ('races_count', models.IntegerField()),
                ('inserted_timestamp', models.DateTimeField(auto_now_add=True)),
            ],
            options={
                'ordering': ('name',),
            },
        ),
        migrations.CreateModel(
            name='Drone',
            fields=[
                ('id', models.AutoField(auto_created=True, primary_key=True, serialize=False, verbose_name='ID')),
                ('name', models.CharField(max_length=250)),
                ('manufacturing_date', models.DateTimeField()),
                ('has_it_competed', models.BooleanField(default=False)),
                ('inserted_timestamp', models.DateTimeField(auto_now_add=True)),
                ('drone_category', models.ForeignKey(on_delete=django.db.models.deletion.CASCADE, related_name='drones', to='drones.DroneCategory')),
            ],
            options={
                'ordering': ('name',),
            },
        ),
        migrations.CreateModel(
            name='Competition',
            fields=[
                ('id', models.AutoField(auto_created=True, primary_key=True, serialize=False, verbose_name='ID')),
                ('distance_in_feet', models.IntegerField()),
                ('distance_achievement_date', models.DateTimeField()),
                ('drone', models.ForeignKey(on_delete=django.db.models.deletion.CASCADE, to='drones.Drone')),
                ('pilot', models.ForeignKey(on_delete=django.db.models.deletion.CASCADE, related_name='competitions', to='drones.Pilot')),
            ],
            options={
                'ordering': ('-distance_in_feet',),
            },
        ),
    ]
